### ISSUE TITLE
"Selected activities" query debugging and validation

### ISSUE THREAD

**Issue by ultimike** [communityinclusion/escoach#39]
@escoach, @mgcobblestone I'm moving this stuff to a new issue because it doesn't belong in an issue titled (and originally meant for) "Dashboard styling" (it's been bugging me for quite some time now). 

Original discussion about this stuff starts around July 12 on #37

**Comment by ultimike** [comment]
Retesting after @mgcobblestone's commit `4035940d2` on August 3.

- Selecting nothing (default values when /dashboard is loaded) results in n=391 (TotalAll). 
- Selecting only "What" = "Getting to know..." results in the "Selected activities" table with n=58 and all zeros for the entire "Selected activities" data row. Something might or might not be wrong here. Debugging on my local shows the following query is used:

```
SELECT
       count(*) AS TotalAll,
       sum(case when ((answer483 in (1) ))  then 1 else 0 end) AS SelectedAll,
       sum(case when ((answer483 in (1) )) AND email = 'alberto.migliore@umb.edu' then 1 else 0 end) AS SelectedMe,
       sum(case when ((answer483 in (1) )) AND provider = 'Institute for Community Inclusion' then 1 else 0 end) AS SelectedProvider,
       sum(case when ((answer483 IS NOT NULL)  AND (email = 'alberto.migliore@umb.edu')) then 1 else 0 end) AS TotalMe,
       sum(case when ((answer483 IS NOT NULL)  AND (provider = 'Institute for Community Inclusion')) then 1 else 0 end) AS TotalProvider
FROM
    surveycampaign_results
WHERE answer483 IN (11658);
```

I'm pretty sure the `(answer483 in (1) )` bits are the issue (I have no idea where the `1` is coming from). Referring back to our logic tree for determining which data table (and therefore query) to use, I believe that this should be the "Selected activities" one which it is, so that's good (I think). 

Taking a step back and thinking about it, while there is a "Selected activity" ("Getting to know"), but no "Who" or "Where" selected, then maybe it makes sense that the "Selected activities" data row is all zeros. @escoach thoughts?

- Selecting "What" = "Getting to know..." and "Who" = "Any interaction" results in n=58 and the rest of the data looks reasonable.
- Selecting "What" = "Getting to know..." and "Who" = "Family, friends.." results in n=58. There's an issue with the query:

```
SELECT
       count(*) AS TotalAll,
       sum(case when ((answer481 in (11641) ))  then 1 else 0 end) AS SelectedAll,
       sum(case when ((answer481 in (11641) )) AND email = 'alberto.migliore@umb.edu' then 1 else 0 end) AS SelectedMe,
       sum(case when ((answer481 in (11641) )) AND provider = 'Institute for Community Inclusion' then 1 else 0 end) AS SelectedProvider,
       sum(case when (((answer481 IS NOT NULL) OR (answer590 IS NOT NULL)) AND email = 'alberto.migliore@umb.edu') then 1 else 0 end) AS TotalMe,
       sum(case when (((answer481 IS NOT NULL) OR (answer590 IS NOT NULL)) AND provider = 'Institute for Community Inclusion') then 1 else 0 end) AS TotalProvider
FROM
    surveycampaign_results
WHERE answer483 IN (11658);
```

@mgcobblestone as you can see, we're not including the "who else" stuff in the three "selected" expressions. Everything else looks good to me. 

- Selecting "What" = "Getting to know..." and "Who" = "Someone from a business" results in n=58. This appears to have the same issue as the previous case (missing "who else"). Fix one, and I think both will be fixed.
- Selecting "What" = "Getting to know..." and "Who" = "Family, friends.. OR Someone from a business" results in n=58. This query suffers from the same ailment as the previous two, with the added bonus of missing a set of "Who" arguments as well ☹️ 

```
SELECT
       count(*) AS TotalAll,
       sum(case when ((answer481 in (11641) ))  then 1 else 0 end) AS SelectedAll,
       sum(case when ((answer481 in (11641) )) AND email = 'alberto.migliore@umb.edu' then 1 else 0 end) AS SelectedMe,
       sum(case when ((answer481 in (11641) )) AND provider = 'Institute for Community Inclusion' then 1 else 0 end) AS SelectedProvider,
       sum(case when (((answer481 IS NOT NULL) OR (answer590 IS NOT NULL)) AND email = 'alberto.migliore@umb.edu') then 1 else 0 end) AS TotalMe,
       sum(case when (((answer481 IS NOT NULL) OR (answer590 IS NOT NULL)) AND provider = 'Institute for Community Inclusion') then 1 else 0 end) AS TotalProvider
FROM
    surveycampaign_results
WHERE answer483 IN (11658);
```

@mgcobblestone as you can see, not only is the "who else" stuff missing, but we're only filtering the expressions by `answer481 in (11641)`. We should be filtering by four things (by my count): `(answer481 in (11641)) OR (answer590 in (12120)) OR (answer481 in (11642)) OR (answer590 in (12119))`. Sound about right?

@mgcobblestone Overall, I think we're on the right track with the refactoring of the "selected activities" query - I know it's a big change; one that would have been nice to avoid. We just need to work through the bugs.

-mike

P.S. @mgcobblestone is there an easy way to output the query used (including the arguments) at the bottom of the data table? It would make things much easier to debug/test. Don't spend more than a few minutes on this - if there's not, no worries.

**Comment by escoach** [comment]
@ultimike wrote: _Taking a step back and thinking about it, while there is a "Selected activity" ("Getting to know"), but no "Who" or "Where" selected, then maybe it makes sense that the "Selected activities" data row is all zeros. @escoach thoughts?_ 

I am not sure I would agree. When "What" = "Getting to know" AND "Who" = NO selection, the table should show 1:16 hrs/day for "All. As it is displayed in the default table (Screenshot below). 

![image](https://user-images.githubusercontent.com/83360935/128186356-e435262f-6529-4a0c-b815-0cfb1ca7959c.png)

**Comment by ultimike** [comment]
@escoach Just to confirm, when "Who" AND "Where" is empty, then it doesn't matter what is selected for the "What". We should just show the default "What summary" (just like when a user visits the /dashboard page initially).

I believe this is what you are saying.

Please confirm.

thanks,  
-mike

**Comment by escoach** [comment]
@ultimike correct.

**Comment by ultimike** [comment]
@mgcobblestone I've updated the [logic wiki doc](https://github.com/communityinclusion/escoach/wiki/Logic-for-determining-which-Dashboard-query-to-use) for determining which query/output to use depending on what is selected based on the previous three comments.

Let me know if you have any questions.

thanks,  
-mike

**Comment by mgcobblestone** [comment]
@ultimike , I've pushed an update that hopefully addresses the main issues here.   I've also added a "debug" checkbox that will display the query and the arguments below the results table when checked.

**Comment by ultimike** [comment]
@mgcobblestone my notes/comments:

1. As @escoach agreed to above, when there is no "Who" or "Where" selected, then the "What" doesn't matter and we should just show the "What summary". I have tweaked the language in the [Logic Wiki page](https://github.com/communityinclusion/escoach/wiki/Logic-for-determining-which-Dashboard-query-to-use) to make this more clear
2. I think the "selected activities" with a selected "What" and "Who" output is looking good. As an example (from my local):

- What = Getting to know, Who = Family, friends, acquaintances of a person I support - Selected all is 2:21.
- What = Getting to know, Who = Someone from a business - Selected all is 1:06.
- What = Getting to know, Who = Family, friends, acquaintances of a person I support OR Someone from a business - Selected all is 3:27. This checks out (1:06 + 2:21).

3. But, the "selected activities" with a selected "What" and "Where" isn't working. Appears to be a PHP Fatal error (<code>Column not found: 1054 Unknown column 'answer0' in 'field list'</code>).

@escoach First off - the "debug" checkbox is temporary while we work through all this debugging of the queries. 

-mike

**Comment by mgcobblestone** [comment]
@ultimike New fix is up.

**Comment by ultimike** [comment]
@mgcobblestone Bad news first ☹️ 

I wasn't clear in my previous comment. When there is no "Who" or "Where" selected, then the "What" doesn't matter and we should just show the "What summary". Currently, if anything is selected for "What" (with "Who" and "Where"), the results are showing the "selected activities" table, not the "What summary". Make sense?

Now, the good news 😄  - it looks like the "selected activities" with a selected "What" and "Where" is now working.

It feels close...

-mike

**Comment by ultimike** [comment]
@mgcobblestone looks good!

@escoach Can you take a look?

thanks,  
-mike

**Comment by escoach** [comment]
@ultimike 
I still see things that are not right.
For example, when I select WHO=Any interaction (No What selected), I get a distribution of time spent with each of the WHO items for a total of 15.9 hours per day for the "All" column (That is the sum of time spent with each of the response items). The total should be 8 hrs/day.

![image](https://user-images.githubusercontent.com/83360935/129052468-443fe3a2-52d4-4e1b-b2e7-403d0aa22982.png)

Similarly, if I select WHAT=Getting to know & WHO= Any interactions, I get 15.5 hrs/day as the total. It should be less than 8 hrs/day. More specifically, it should be the amount of time spent just for getting to know job seekers (Probably 1 or 2 hours per day)

![image](https://user-images.githubusercontent.com/83360935/129052908-2deabca0-a9a4-48b2-83a9-9211903c3292.png)

Also, when I select "WHAT=Getting to know", (No WHO and no WHERE selected) I get 8 hours per day regardless of the group (Me, Team, All). 

![image](https://user-images.githubusercontent.com/83360935/129055645-27a04a4f-4511-4cec-9482-c5bc337f347f.png)

**Comment by escoach** [comment]
@ultimike 
Excellent points. In the future, I will make sure to remember those considerations. Not sure why I completely overlooked the "Who else" and the combined metrics. Thanks.
Also, I  tried again the "What" = "Getting to know" and no "Who" or "Where" selected and looks fine. NOt sure why yesterday I got a questionable result. 
The bottom line is that this looks good!

**Comment by ultimike** [comment]
@escoach That's promising news. Are you saying that the data is good-to-go for the up-to-date timeframe?

@mgcobblestone I think we need to now turn our attention to getting the trends queries working again as they're not returning any table data 😱 

-mike

**Comment by mgcobblestone** [comment]
@ultimike I am seeing results for trends queries on my local.  I just pushed up updates to the templates to add the debugging to the trends template.  Let me know if that provides any useful info.

**Comment by ultimike** [comment]
No - something is definitely going on. I just tried it on the [Dev site](http://sp-escoach.communityinclusion.org/dashboard) and I get nothing - not even a query to debug...

I selected only "What" = "Getting to know..." and "Time frame" = "Quarterly trends".

Report log has: 

```
TypeError: Argument 1 passed to Drupal\survey_dashboard\Query\BaseQuery::buildSelectedSums() must be of the type array, 
null given, called in  /home/ici/escoach.communityinclusion.org/escoach/web/modules/custom/survey_dashboard/src/Service/QueryBuilder.php 
on line 493 in Drupal\survey_dashboard\Query\BaseQuery->buildSelectedSums() (line 355 of 
/home/ici/escoach.communityinclusion.org/escoach/web/modules/custom/survey_dashboard/src/Query/BaseQuery.php)
```

-mike

**Comment by escoach** [comment]
@ultimike 
I am not sure what's going with my computer. But I keep getting strange outputs. For example, here is the summary screenshot with nothing selected:

![image](https://user-images.githubusercontent.com/83360935/129273842-a585287d-450a-46ab-a6e3-e2578caf561c.png)

Now I select What = "Observing a job seeker" , hit submit, and nothing changes (Screenshot below). I restarted my computer, and i switched from Crome to Firefox. But I got the same results. I will check again tomorrow (Friday) with a colleague on his computer. Thanks.  

![image](https://user-images.githubusercontent.com/83360935/129273914-51f0a735-ab22-49da-9986-0a7f6c19cf29.png)

**Comment by ultimike** [comment]
@escoach This is the correct behavior. From this thread 9 days ago:

> @escoach Just to confirm, when "Who" AND "Where" is empty, then it doesn't matter what is selected for the "What". We should just show the default "What summary" (just like when a user visits the /dashboard page initially).

> @ultimike correct.

-mike

**Comment by escoach** [comment]
@ultimike I am not sure why I responded that way. It's not correct. Here is how we intended it to be: https://github.com/communityinclusion/escoach/issues/18

How big of a problem is to fix this?

**Comment by escoach** [comment]
@ultimike a colleague pointed out the following discrepancy:

"What = Informational interviews" AND "Where = Anyplace" yields the following:
2 hrs/day at a business  + 2 hrs/day at a residence + 4 hrs/day in other community settings = **8 hrs/day**. 
 (See screenshot below, column "All")
However, the total should be less than 1:16 hrs/day. This is because "Informational interviews" is a subset of "Getting to know", which takes only 1:16 hrs/day (See summary table). 


![image](https://user-images.githubusercontent.com/83360935/129396101-07dc9ab3-a97b-4a29-a661-f76d273fa0ba.png)
 
Another discrepancy is when the column "All" shows data whereas the columns "Me" and "My team" do not. For example, in the screenshot above, first row, column "All" shows 2 hrs/day at a business. However, columns "Me" and "My Team" show zero hours. Clearly, if there is a value in the "All" column, there must be a value in at least "Me". the "My team" column can be empty if a user is there alone, without being part of a team. I understand that this might be just because of the way the data in the dataset are (dis)organized. Or is it a real issue?

**Comment by ultimike** [comment]
@escoach,

I have reviewed the case where "What" = "Informational interview or a business tour and"Where" = "Anywhere". This is what I found:

- The query is correct. For example, it is reporting that of the 4 responses that selected the following, only 1 response indicated "Where" = "At a business (Not provider-owned)".
  - "What" = "Informational interview or a business tour" AND
  - "What" (top-level) is not null
- We properly applied the equation you provided (BusinessAll/TotalAll*8/24) in order to figure out the hours/day. Looking at the same example as above: Hours/day for "At a business (Not provider-owned)" is equal to: 1/4*8/24 = 2 hours/day.

Regarding the second discrepancy that you're seeing, I don't follow your logic. Why can't there be results that have been submitted that weren't by you or your team? If someone else from some other team has a result, then it will be shown. 

The "Me" and "My Team" data is, by definition, subsets of "All". I would argue the exact opposite of what you're saying. If either "Me" or "My team" has data, then "All" cannot be empty.

-mike

**Comment by ultimike** [comment]
> @ultimike I am not sure why I responded that way. It's not correct. Here is how we intended it to be: #18
> 
> How big of a problem is to fix this?

I'm not sure, but this is getting quite frustrating with time getting short.

I would ask that we leave this alone for now so that we can confirm that what we currently have is working. 

-mike

**Comment by escoach** [comment]
@ultimike Thank you for your comments. 

You wrote: _"...We properly applied the equation you provided (BusinessAll/TotalAll(8/24)) in order to figure out the hours/day. Looking at the same example as above: Hours/day for "At a business (Not provider-owned)" is equal to: 1/4(8/24) = 2 hours/day...."_ 

Thank you. However, the "TotalAll" should be 361 (Total responses for "All"), not 4. 

Therefore: (1/361) x 8 = .022 hrs/day that is 1.3 minutes / day. 

I checked the hrs/day for each of the "What" (ID483) response items (Screenshot below) and they match the figures shown on the dashboard. So it seems that the issue manifests only when "What" is combined with "Where" or "Who".

![image](https://user-images.githubusercontent.com/83360935/129497596-d4f83cb3-5437-4502-ab5d-cab0ef72116a.png)

As for the discrepancy, you wrote: _"....Regarding the second discrepancy that you're seeing, I don't follow your logic. Why can't there be results that have been submitted that weren't by you or your team? If someone else from some other team has a result, then it will be shown....."_

I agree with you. Please discard my earlier comment about this issue.


PS: I honestly do not remember if I provided the formula that you wrote above. It was a long time ago when we discussed it. But if I did, it is not working. I am glad we caught this issue.

**Comment by ultimike** [comment]
@escoach,

The equation that we've been using all along is from the "Explanation_survey_calc.docx"
provided to us at the onset of this project. The extra "/24" I had in my previous
comment can be ignored. I use that in the Google Sheet in order to convert things
properly for testing.

You wrote:

> I checked the hrs/day for each of the "What" (ID483) response items (Screenshot
below) and they match the figures shown on the dashboard. So it seems that the
issue manifests only when "What" is combined with "Where" or "Who".

I _think_ what you're suggesting is that "Total respondents (all)" should always
be the total number of people who submitted the survey (perhaps with a non-empty
"What"). If we do that, I'm pretty sure then none of the columns will add up to
8 or 16 hours. For example, if we use my example from my previous comment:

TotalAll: 391
TotalMe: 134
TotalProvider: 68

Using the following example
equation: `1/391 x 8 = 0.020 hours/day = 1.2 minutes/day = 0:01`
This results in:

Informational interview...   | All hrs/day | Me hrs/day | Provider hrs/day
--- | --- | --- | ---
At a business (Not provider-owned) | 0:01  | 0 | 0
At the residence of a person I support | 0:01  | 0 | 0
Residence/business/community |  0:05 | 0:07 | 0
Vehicle | 0  | 0 | 0
Office | 0  | 0 | 0
Office/residence/community | 0:04  | 0:07 | 0
Community | 0:02  | 0:07 | 0
Other | 0  | 0 | 0
Total respondents | 391 | 134 | 68

I was using your account as "Me" and "Institute for Community Inclusion" as the
provider, but since you weren't associated with the ICI provider when you
submitted the surveys, none of them are tallied in this example.

Please review these results and this logic - I don't want to have to ask
@mgcobblestone to have to change the queries again until we are certain we
know what we're doing.

@mgcobblestone - if all this looks good, then I think the easiest way to do this
be be by adding a second query for the 3 "total" values for only the "selected
activities" case. We can discuss once we hear back from @escoach.

thanks,
-mike

**Comment by escoach** [comment]
@ultimike 
It looks like we are in agreement on the equation: I replicated your computations for the "All" column in the table that you shared and I got about the same results (See my table below). Well, I got only 3 minutes for office/residence/community whereas your table shows 4 minutes. It must be some rounding issues. Also, your TotalAll is 391 whereas I have 361.

Note that the "n" to be shown at the bottom of the table on the dashboard should be 4 (The numerator), not 391 (The denominator). We want the users to be aware of how small the "n" can get for certain metrics.

![image](https://user-images.githubusercontent.com/83360935/129621339-1c860121-2da2-4860-8269-dcd789e05442.png)

Here are the equations (n/N*8): 
In my office or home office: 1/361*8 = 0:01 min
At the residence of a person I support: 0/361*8 = 0:00 min
At a business (Not provider-owned): 1/361*8 = 0:01 min
In other community integrated settings: 2/361*8 = 0:02 min
[...]

Having said that, note that my Total Ns do not match with yours. I am using the file "surveycampaign_results.csv".

- TotalAll: you have 391 whereas I have only 361. I counted ID483>1 i.e. "What" is not empty.
- TotalMe: you have 134 whereas I have 110. I counted "email" = "alberto.migliore@umb.edu" AND "What" is not empty.
- TotalProvider: you have 68 whereas I have 38. I counted "provider" = "Institute for Community Inclusion" AND "What" is not empty.

I hope this helps to clarify and next steps.
Let me know how I can help
Alberto

**Comment by ultimike** [comment]
Do we need to add a new "n" column?

-mike

**Comment by escoach** [comment]
No. I added it in my table just for context. I am afraid it would make things too complicated.

**Comment by ultimike** [comment]
@escoach I'm using a slightly different (newer) dataset on my local that you are on the         http://sp-escoach.communityinclusion.org/  site - this is why our numbers are slightly different.

I believe (hope) I understand the changes that you are asking for. 

- The "Total respondents" row data that is displayed should not change.  It should continue to come from the current query.
- The current query we are using to tally the number of responses does not need to change, only the way we calculate the values in the table.
- We need to use a different query to calculate the denominator for the hrs/day (and hrs/week) calculations - one where we get the total number of responses without filtering by the selected "WHAT". 

@escoach now the big question - when do we apply this new logic? Only when "Where" = "Anyplace" or "Who" = "Any interaction"?

I feel like every time we have a handle on things, we have to make some big change (like this) that can potentially break something else. 

-mike

@mgcobblestone Don't do anything yet, we really need to make sure we understand the scope of the changes here, but here's an example of a "second query" I mentioned above:

```
SELECT
    count(*) AS AllResponses,
    sum(case when ((answer482 IS NOT NULL) AND (email = 'alberto.migliore@umb.edu')) then 1 else 0 end) AS MeResponses,
    sum(case when ((answer482 IS NOT NULL) AND (provider = 'Institute for Community Inclusion')) then 1 else 0 end) AS ProviderResponses
FROM surveycampaign_results
WHERE answer483 IS NOT NULL;
```

Notice that it is similar to the expressions for `TotalAll`, `TotalMe`, and `TotalProvider` but without the filter in the `WHERE` clause.

-mike

**Comment by escoach** [comment]
@ultimike 
See my responses after each of your statements 

You wrote:_- The "Total respondents" row data that is displayed should not change. It should continue to come from the current query._

Alberto: The "**Total respondents**" row is the numerator in the equation that was provided in "Explanation_survey_calc.docx":
Number of Hrs/Day = (**Number of responses to a specific question** / Total number of responses) x 8.
Specifically, "Number of responses to a specific question" = numerator = "Total respondents" in the row at the bottom of the table on the dashboard.

For example, applying this concept to the example discussed yesterday, here is what the equation and the "Total respondents" look like for "All":
Hrs/day "Getting to know = Informational interview" AND "Where= At a business" = 1 / 361 x 8 = 0:01 minutes/day
The "Total respondents" displayed at the bottom of the table is "1", showing that doing an informational interview at a business was reported only once. 
 
You wrote: _- The current query we are using to tally the number of responses does not need to change, only the way we calculate the values in the table._

Alberto: I hesitate to confirm because I am not sure I know exactly what you are referring to. I hope that my response to your first question helps to guide you on what changes are needed if needed. Or we can chat tomorrow.

_- We need to use a different query to calculate the denominator for the hrs/day (and hrs/week) calculations - one where we get the total number of responses without filtering by the selected "WHAT"._

Alberto: I  am not sure I fully understand the question without more context. 


You wrote: - _now the big question - when do we apply this new logic? Only when "Where" = "Anyplace" or "Who" = "Any interaction"?_

Alberto: My sense is that examples 1-4 in the file "Es_coach_dashboard_mockup.docx" that Paul shared at the onset of the project are still relevant for what we are trying to achieve. But I am happy to talk more tomorrow if I am missing something.


@thinkcollege feel free to add comments if needed.

**Comment by ultimike** [comment]
@escoach Let's close this in lieu of #43 and other future (more focused) tasks. Cool with you?

-mike

**Comment by ultimike** [comment]
Closing.
